File: c_code/19计科huster的第一篇题解.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */


struct ListNode* reverseKGroup(struct ListNode* head, int k){
    struct ListNode *headp = head, *temp, *head0;
    head0 = head;
    int i;
    for(temp = headp, i = 0; temp && i < k; i++, temp = temp->next)
    ;                               //作用是确定出一个完整的反转小组
    if(!temp && i < k)
        return headp;               //若不足k元,直接返回原本头节点
    struct ListNode *cur, *newhead = temp;
    while(headp != temp)
    {
        cur = headp;
        headp = headp->next;
        cur->next = newhead;
        newhead = cur;
    }                               //“伪造”的递归核心,由迭代实现每个小组的反转
    head0->next = reverseKGroup(temp, k);
    return newhead;                 //返回该小组的“头节点”作为上一小组“尾节点”后继
}

File: output/19计科huster的第一篇题解/19计科huster的第一篇题解.rs
#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode {
            next: None,
            val,
        }
    }
}

pub fn reverse_k_group(head: Option<Box<ListNode>>, k: i32) -> Option<Box<ListNode>> {
    if head.is_none() || k <= 1 {
        return head;
    }
    
    // Check if we have at least k nodes
    let mut temp = head.as_ref();
    for _ in 0..k {
        if temp.is_none() {
            return head; // Not enough nodes, return original head
        }
        temp = temp.unwrap().next.as_ref();
    }
    
    // Clone the remaining part for recursion
    let remaining = if let Some(node) = temp {
        Some(node.clone())
    } else {
        None
    };
    
    // We have at least k nodes, proceed with reversal
    let mut current = head;
    let mut prev = reverse_k_group(remaining, k);
    
    // Reverse k nodes
    for _ in 0..k {
        if let Some(mut node) = current {
            current = node.next.take();
            node.next = prev;
            prev = Some(node);
        }
    }
    
    prev
}